spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageResource.java

```diff
@@ -0,0 +1,332 @@
+	@Override
+	public boolean isOpen() {
+		return false;
+	}
+
+	/**
+	 * Since the gs: protocol will normally not have a URL stream handler registered,
+	 * this method will always throw a {@link java.net.MalformedURLException}.
+	 * @return The URL for the GCS resource, if a URL stream handler is registered for the gs protocol.
+	 */
+	@Override
+	public URL getURL() throws IOException {
+		return this.location.toURL();
+	}
+
+	@Override
+	public URI getURI() throws IOException {
+		return this.location;
+	}
+
+	/**
+	 * Gets the underlying storage object in Google Cloud Storage.
+	 * @return The storage object, will be null if it does not exist in Google Cloud Storage.
+	 * @throws IOException
+	 * @throws IllegalStateException if the resource reference is to a bucket, and not a blob.
+	 */
+	public Blob getBlob() throws IOException {
+		return this.storage.get(getBlobId());
+	}
+
+	/**
+	 * Creates a signed URL to an object, if it exists. This method will fail if this storage
+	 * resource was not created using service account credentials.
+	 * @param timeUnit the time unit used to determine how long the URL is valid.
+	 * @param timePeriods the number of periods to determine how long the URL is valid.
+	 * @return the URL if the object exists, and null if it does not.
+	 * @throws IllegalStateException if the resource reference is to a bucket, and not a blob.
+	 */
+	public URL createSignedUrl(TimeUnit timeUnit, long timePeriods)  {
+		if (LOGGER.isWarnEnabled()) {
+			if (!exists()) {
+				LOGGER.warn("Creating signed URL for non-existing GCS object " + this.location);
+			}
+		}
+
+		return this.storage.signUrl(
+				BlobInfo.newBuilder(getBlobId()).build(), timePeriods, timeUnit);
+	}
+
+	/**
+	 * Creates the blob that this {@link GoogleStorageResource} represents in Google Cloud
+	 * Storage.
+	 * @return the created blob object
+	 * @throws StorageException if any errors during blob creation arise,
+	 * such as if the blob already exists
+	 * @throws IllegalStateException if the resource reference is to a bucket, and not a blob.
+	 */
+	public Blob createBlob() throws StorageException {
+		return this.storage.create(BlobInfo.newBuilder(getBlobId()).build());
+	}
+
+	/**
+	 * Creates the bucket that this resource references in Google Cloud Storage.
+	 * @return the {@link Bucket} object for the bucket
+	 * @throws StorageException if any errors during bucket creation arise,
+	 * such as if the bucket already exists
+	 */
+	public Bucket createBucket() {
+		return this.storage.create(BucketInfo.newBuilder(this.bucketName).build());
+	}
+
+	/**
+	 * Returns the {@link Bucket} associated with the resource.
+	 * @return the bucket if it exists, or null otherwise
+	 */
+	public Bucket getBucket() {
+		return this.storage.get(this.bucketName);
+	}
+
+	/**
+	 * Checks for the existence of the {@link Bucket} associated with the resource.
+	 * @return true if the bucket exists
+	 */
+	public boolean bucketExists() {
+		return getBucket() != null;
```

Comment:
Why do we need this extra method and why is it not proxying `exists()`?

Reply:
It's possible for the bucket to exist, but the file not to exists. See tests.

Reply:
This bucket can either represent a bucket or a blob.
If it represents a blob, there's no point in supporting a `bucketExists()` method.
This is the issue of having this object represent two different things.

Perhaps we shouldn't consider a bucket as a `Resource`, since you can't read it and, arguably, can't write to it (creating files isn't really writing to a bucket..)?